typing/typeopt.ml

```diff
@@ -61,17 +58,18 @@ let is_base_type env ty base_ty_path =
   | Tconstr(p, _, _) -> Path.same p base_ty_path
   | _ -> false
 
-let is_immediate = function
-  | Type_immediacy.Unknown -> false
-  | Type_immediacy.Always -> true
-  | Type_immediacy.Always_on_64bits ->
-      (* In bytecode, we don't know at compile time whether we are
-         targeting 32 or 64 bits. *)
-      !Clflags.native_code && Sys.word_size = 64
+let is_immediate env ty =
+  let imm : Type_immediacy.t =
+    (* In bytecode, we don't know at compile time whether we are
+       targeting 32 or 64 bits. *)
+    if !Clflags.native_code && Sys.word_size = 64 then Always_on_64bits
+    else Always
+  in
+  Result.is_ok (Ctype.check_type_immediate env ty imm)
```

Comment:
I find the new definition harder to read than before. It comes from previous design iterations of this PR, which had functions to check whether a given type satisfied a given immediacy, but no function to actually compute the immediacy of a type. But this is not the case anymore. Could we write the following?

```ocaml
let is_immediate env ty =
  match Ctype.(type_immediacy_head env ty) with
  ...
```

?

(I think this would require tuning `type_immediacy` to work correctly on unboxed types, as suggested in another comment.)

Reply:
Interesting!  I found the old version confusing (particularly, `is_immediate` seemed an odd name for a `Type_immediacy.t -> bool`).

I think we still do not have a function to calculate the exact immediacy - as described above the issue is not just unboxed types but also the approximation in the kind for abbreviations.  We could add such a function - an analogue of `check_type_immediate` that returns the immediacy.  My instinct is that this would be unnecessarily duplicative if it's only used in this one place, but I'm open to other opinions.